src/librbd/ImageState.cc

```diff
@@ -22,16 +23,231 @@ namespace librbd {
 using util::create_async_context_callback;
 using util::create_context_callback;
 
+class ImageUpdateWatchers {
+public:
+
+  ImageUpdateWatchers(CephContext *cct) : m_cct(cct),
+    m_lock(util::unique_lock_name("librbd::ImageUpdateWatchers::m_lock", this)) {
+  }
+
+  ~ImageUpdateWatchers() {
+    assert(m_watchers.empty());
+    assert(m_in_flight.empty());
+    assert(m_pending_unregister.empty());
+    assert(m_on_flush_finish == nullptr);
+    assert(m_on_shut_down_finish == nullptr);
+
+    destroy_work_queue();
+  }
+
+  void flush(Context *on_finish) {
+    ldout(m_cct, 20) << "ImageUpdateWatchers::" << __func__ << dendl;
+    {
+      Mutex::Locker locker(m_lock);
+      assert(m_on_flush_finish == nullptr);
```

Comment:
Not sure this is a safe assumption -- can get repeated notifications and/or races with multiple updates.